File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
    queue = deque()
    n = len(maps)
    m = len(maps[0])
    visited = [[False] * m for _ in range(n)]
    x,y=0,0
    visited[0][0] = True
    queue.append((x,y))
    
    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]
            if 0 <= nx < n and 0 <= ny < m and maps[nx][ny] != 0:
                if not visited[nx][ny]:
                    maps[nx][ny] = maps[x][y] + 1
                    queue.append((nx, ny))
                    visited[nx][ny] = True
    if maps[n-1][m-1] == 1: return -1
    else: return maps[n-1][m-1]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
using namespace std;

int dx[4] = {1,-1,0,0};
int dy[4] = {0,0,1,-1};

int bfs(int starty, int startx, vector<vector<int>>& maps){
    int n = maps.size(); int m = maps[0].size();
    queue<pair<int,int>> q;
    vector<vector<int>> dist(n, vector<int>(m,-1));
    
    q.push({starty,startx});
    dist[starty][startx] = 1;
    
    while(!q.empty()){
        auto[y,x] = q.front(); q.pop();
        for(int i = 0; i < 4; i++){
            int ny = dy[i] + y;
            int nx = dx[i] + x;
            
            if(ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
            if(maps[ny][nx] == 0) continue;
            if(dist[ny][nx] != -1) continue;
            
            dist[ny][nx] = dist[y][x] + 1;
            q.push({ny,nx});
        }
        
    }
    
    return dist[n-1][m-1];
}

int solution(vector<vector<int>> maps)
{
    return bfs(0,0,maps);
}